### ISSUE TITLE
Add encode/decode logic for ML-KEM

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#25885]
This issue is to track standards for persisting ML-KEM keys as well as to suggest implementing them. In terms of implementation, this issue thus suggests adding encoding/decoding logic for all ML-KEM key components.

This below is to track the standards defining these formats to enable subsequent implementation of those (draft only):

- Standard for public key format: "encapsulation key ek" binary format from [FIPS203](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.pdf) section 6.1 and encoded as per draft specification https://datatracker.ietf.org/doc/draft-ietf-lamps-kyber-certificates/06/
- Standard for private key format: "seed"-format as per section 5 in draft specification https://datatracker.ietf.org/doc/draft-ietf-lamps-kyber-certificates/06/

Proposed target for any PRs currently is the "ml-kem" feature branch.

Input solicited, particularly from @tomato42 and @t-j-h

**Comment by tomato42** [comment]
I see both specified in the [draft-ietf-lamps-kyber-certificates-06](https://datatracker.ietf.org/doc/draft-ietf-lamps-kyber-certificates/06/) document. It already references the official NIST OIDs.

**Comment by baentsch** [comment]
> I see both private and public key format specified in the [draft-ietf-lamps-kyber-certificates-06](https://datatracker.ietf.org/doc/draft-ietf-lamps-kyber-certificates/06/) document. It already references the official NIST OIDs.

Thanks for that pointer, @tomato42 . So your request is to implement this draft standard as-is? The public key format I think is pretty much undisputed -- so added it above. Storing the private key as "seed" only seems to not be generally agreed -- can you point to tooling that could be used to ascertain interoperability of any possible implementation?

**Comment by tomato42** [comment]
Having attended the LAMPS session on this IETF this week, there is very strong consensus for storing private keys as seeds.

for tooling, I think this new stuff is already in the https://github.com/IETF-Hackathon/pqc-certificates/ repo...

**Comment by baentsch** [comment]
Thanks for that "live IETF" report, @tomato42 ! 

> yes, I think we can implement them

Does this indicate your willingness to take on this issue and work towards a PR for the "ml-kem" feature branch then?

**Comment by tomato42** [comment]
no, it means willingness to test it for interoperability with itself and with the implementation I still need to implement in https://github.com/GiacomoPope/kyber-py/pull/80 (it can do past oqsprovider behaviour, need to update it to do the LAMPS seed strorage)

**Comment by andrewkdinh** [comment]
@vdukhovni can this issue be closed now?

**Comment by t-j-h** [comment]
Closing issue as we have encoders/decoders in place and follow up PRs being tracked.
